tunnel/intra/doh/client_auth.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2020 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package doh
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"io"
+	"sync"
+)
+
+// CertificateLoader interface for requesting ClientAuth instances.
+type CertificateLoader interface {
+	// Request a ClientAuth instance (blocking).
+	// returns nil (no authentication) or a ClientAuth instance.
+	LoadClientCertificate() ClientAuth
+}
+
+// ClientAuth interface for providing TLS certificates and signatures.
+type ClientAuth interface {
+	// GetClientCertificate returns the client certificate (if any).
```

Comment:
Why would this ever return nil?  Wouldn't it make more sense for the CertificateLoader to return nil in that case?

Reply:
This should never return nil, as we attempt to set signer (the ClientAuth instance) to nil in that case.  If GetClientCertificate returns nil then the ClientAuth instance has changed it's mind about whether to provide certificates.  I don't anticipate that this will happen, but if it does we handle it.

Reply:
OK.  It sounds like returning nil would be an error condition, so please log it.

Reply:
ClientAuth.GetClientCertificate() returning nil will be logged in every instance now.